### ISSUE TITLE
`resume_from_checkpoint` loses the epoch seed: the rebuilt sampler replays the epoch-0 order, so already-trained samples are re-trained

### ISSUE THREAD

**Issue by Si1w** [modelscope/ms-swift#10050]
### Checklist / 检查清单

- [x] I have searched existing issues, and this is a new bug report. / 我已经搜索过现有的 issues，确认这是一个新的 bug report。

### Bug Description / Bug 描述

When training resumes from a checkpoint that lies inside epoch index `>= 1` (i.e. `num_train_epochs > 1` and the checkpoint is not in the first epoch), the rest of that epoch is trained on the **wrong data order**.

`BatchSamplerShard` derives the per-epoch permutation from `data_seed + epoch` via `set_epoch()`. On resume, HF `Trainer` calls `epoch_dataloader.set_epoch(epoch)` and then `skip_first_batches(...)`. ms-swift patches `skip_first_batches` so that it rebuilds the dataloader through `get_train_dataloader(skip_batches=...)`:

```python
# swift/trainers/mixin.py
@contextmanager
def _patch_skip_first_batches(self):
    from transformers import trainer
    origin_skip_first_batches = trainer.skip_first_batches

    def skip_first_batches(dataloader, num_batches=0):
        if isinstance(dataloader, (DataLoaderShard, DataLoaderDispatcher)):
            # DataLoaderMixin
            return self.get_train_dataloader(skip_batches=num_batches)
        else:
            return origin_skip_first_batches(dataloader, num_batches)
```

`get_train_dataloader` constructs a **new** `BatchSamplerShard`, whose `curr_seed` is reset to `base_seed` (= `data_seed`). Nobody calls `set_epoch()` on the new sampler, so the resumed dataloader iterates the **epoch-0 permutation** and skips its first `N` batches, instead of the permutation of the epoch that was actually in progress.

Consequences for the remainder of the resumed epoch:

- most of the samples it trains on were already trained earlier in the same epoch (they were in the prefix of the correct permutation);
- the samples that should have been trained (the tail of the correct permutation) are never seen in that epoch.

The bug is invisible when resuming inside epoch index 0, because there `data_seed + 0 == data_seed`. This is why e.g. #6049 (resume at epoch 0.28) reproduces the loss exactly.

This is the same `set_epoch` semantics that #9685 and #9902 already fixed for other paths: #9685 makes `DataLoaderShard.set_epoch` propagate through a wrapping `batch_sampler.batch_sampler` (the `SkipBatchSampler` layout accelerate's original `skip_first_batches` produces), and #9902 adds `set_epoch` to `DataLoaderDispatcher` so that streaming datasets do not iterate in the same order every epoch. The patched `skip_first_batches` bypasses both, because it never calls `set_epoch` on the dataloader it builds.

### What we observed in a real run

Qwen3-8B full-parameter SFT, 2 x H200, `deepspeed zero3`, `num_train_epochs 4`, `data_seed 42`, 826 optimizer steps per epoch. The run crashed twice and was resumed from `checkpoint-1500` (epoch 1.82) and `checkpoint-3000` (epoch 3.63).

After the first resume, global steps 1501–1586 were replayed from the same weights. If the data order were preserved the two passes should give near-identical per-step losses (as in #6049). Instead:

| Comparison (same weights, same global steps) | first pass | replay | ratio |
|---|---|---|---|
| mean train loss over 86 steps | 0.1637 | 0.0791 | 0.48 |
| per-step Pearson correlation of the two passes | | | -0.23 |
| mean loss 50 steps before / after the second resume | 0.0199 | 0.0100 | 0.50 |

The correlation near zero shows the batches differ; the halved loss is the model re-seeing samples it already trained on in that epoch. A simulation with our corpus size gives 81% duplicated samples (and 15% of the corpus never trained in that epoch) for the first resume, 64% / 24% for the second.

### How to Reproduce / 如何复现

Self-contained script; it only uses `DataLoaderMixin` and `_patch_skip_first_batches`, in the same order HF `Trainer._inner_training_loop` calls them (`set_epoch` at trainer.py L2581, `skip_first_batches` at L2600 in transformers 4.57.6).

```python
from types import SimpleNamespace
from swift.trainers.mixin import SwiftMixin, DataLoaderMixin
from transformers import trainer as hf_trainer

dummy = SimpleNamespace(
    template=SimpleNamespace(sequence_parallel_size=1),
    train_dataset=list(range(1000)),
    _train_batch_size=1,
    data_collator=lambda batch: batch,
    accelerator=SimpleNamespace(device=None),
    args=SimpleNamespace(
        dataloader_num_workers=0, dataloader_pin_memory=False,
        dataloader_persistent_workers=False, dataloader_prefetch_factor=None,
        dataloader_drop_last=False, train_dataloader_shuffle=True, data_seed=42,
        deepspeed=None, group_by_length=False, process_index=0),
)
dummy.get_train_dataloader = DataLoaderMixin.get_train_dataloader.__get__(dummy)
dummy._patch_skip_first_batches = SwiftMixin._patch_skip_first_batches.__get__(dummy)
order = lambda dl: [b[0] for b in dl]

# fresh run, epoch index 1
dl = dummy.get_train_dataloader(); dl.set_epoch(1); epoch1 = order(dl)
# fresh run, epoch index 0
dl = dummy.get_train_dataloader(); dl.set_epoch(0); epoch0 = order(dl)

# resume inside epoch index 1 after 600 batches
dl = dummy.get_train_dataloader(); dl.set_epoch(1)
with dummy._patch_skip_first_batches():
    resumed = hf_trainer.skip_first_batches(dl, 600)
tail = order(resumed)

print("tail == epoch1[600:] ?", tail == epoch1[600:])   # expected True, got False
print("tail == epoch0[600:] ?", tail == epoch0[600:])   # expected False, got True
print("curr_seed:", resumed.batch_sampler.batch_sampler.curr_seed)  # 42, expected 43
print("already trained this epoch:", len(set(tail) & set(epoch1[:600])) / len(tail))  # 0.59
```

Output:

```
tail == epoch1[600:] ? False
tail == epoch0[600:] ? True
curr_seed: 42
already trained this epoch: 0.59
```

### Additional Information / 补充信息

- ms-swift 4.5.2 (the code is unchanged on `main` as of 2026-09-06)
- transformers 4.57.6, accelerate 1.14.0, torch 2.8.0+cu128, deepspeed zero3
- Qwen3-8B, `--tuner_type full`, `--dataset_shuffle true` (default), `--data_seed 42` (default), `--dataloader_num_workers 2`
- Linux, 2 x NVIDIA H200

```python
def skip_first_batches(dataloader, num_batches=0):
    if isinstance(dataloader, (DataLoaderShard, DataLoaderDispatcher)):
        new_dataloader = self.get_train_dataloader(skip_batches=num_batches)
        old_sampler = getattr(dataloader, 'batch_sampler', None)
        if hasattr(old_sampler, 'curr_seed') and hasattr(new_dataloader, 'set_epoch'):
            new_dataloader.set_epoch(old_sampler.curr_seed - old_sampler.base_seed)
        return new_dataloader
    return origin_skip_first_batches(dataloader, num_batches)
```
